massa-pos-worker/src/worker.rs

```diff
@@ -61,33 +58,38 @@ impl SelectorThread {
                 initial_rolls,
             };
 
-            this.run(stop_flag)
+            this.run()
         })
     }
 
-    /// Check if cycle info changed or new and compute the draws
-    /// for future cycle.
-    /// # Arguments
-    /// * `cycle_info`: a cycle info with roll counts, seed, etc...
-    fn run(mut self, stop_flag: Arc<AtomicBool>) -> PosResult<()> {
+    /// Thread loop.
+    ///
+    /// While a `Stop` command isn't sent, pop `input_data` and compute
+    /// draws for future cycle.
+    fn run(mut self) -> PosResult<()> {
         loop {
-            if stop_flag.load(std::sync::atomic::Ordering::Relaxed) {
-                break;
-            }
-            let input_data = self.input_data.clone();
-            let mut data = input_data.1.lock();
-            if let Some(cycle_info) = data.pop_front() {
-                self.draws(cycle_info)?
+            let cycle_info = {
+                let mut data = self.input_data.1.lock();
+                match data.pop_front() {
+                    Some(Command::CycleInfo(cycle_info)) => Some(cycle_info),
+                    Some(Command::Stop) => break,
+                    None => None,
+                }
+            };
+
+            if let Some(cycle_info) = cycle_info {
+                self.draws(cycle_info)?;
             }
+
             // Wait to be notified of new input
             // The return value is ignored because we don't care what woke up the condition variable.
-            self.input_data.0.wait(&mut data);
+            self.input_data.0.wait(&mut self.input_data.1.lock());
```

Comment:
This in fact creates a second critical section, separate from the one above. It doesn't actually matter much in this particular case, and it's still better to consolidate the logic into one. 

I think the below run loop is much clearer:

```rust
    fn run(mut self) -> PosResult<()> {
        loop {
            let cycle_info = {
                let mut data = self.input_data.1.lock();
                loop {
                    match data.pop_front() {
                        Some(Command::CycleInfo(cycle_info)) => break cycle_info,
                        Some(Command::Stop) => return Ok(()),
                        None => self.input_data.0.wait(&mut data),
                    }
                }
            };
            self.draws(cycle_info)?;
        }
    }
```

Reply:
Great! Very nice :+1: